src/ResourceManager/Resources/Commands.Resources/Models.ProviderFeatures/ProviderFeatureClient.cs

```diff
@@ -0,0 +1,153 @@
+        /// </summary>
+        public const string RegisteredStateName = "Registered";
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ProviderFeatureClient"/> class.
+        /// </summary>
+        public ProviderFeatureClient()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ProviderFeatureClient"/> class.
+        /// </summary>
+        /// <param name="context">The azure context</param>
+        public ProviderFeatureClient(AzureContext context)
+        {
+            this.FeaturesManagementClient = AzureSession.ClientFactory.CreateClient<Management.Resources.FeatureClient>(context, AzureEnvironment.Endpoint.ResourceManager);
+        }
+
+        /// <summary>
+        /// The features management client
+        /// </summary>
+        public IFeatureClient FeaturesManagementClient { get; set; }
+
+
+        /// <summary>
+        /// Lists the features that ARM knows about
+        /// </summary>
+        /// <param name="resourceProviderNamespace">When specified, returns all features that are defined by this resource provider namespace</param>
+        /// <param name="featureName">When specified, returns all features that have this name</param>
+        public virtual PSProviderFeature[] ListPSProviderFeatures(string resourceProviderNamespace = null, string featureName = null)
+        {
+            return this.ListPSProviderFeatures(resourceProviderNamespace: resourceProviderNamespace, featureName: featureName, listAvailable: false);
+        }
+
+        /// <summary>
+        /// Lists the features that ARM knows about
+        /// </summary>
+        /// <param name="resourceProviderNamespace">When specified, returns all features that are defined by this resource provider namespace</param>
+        /// <param name="listAvailable">When set to true, lists all features that are available including those not registered on the current subscription</param>
+        public virtual PSProviderFeature[] ListPSProviderFeatures(bool listAvailable, string resourceProviderNamespace = null)
+        {
+            return this.ListPSProviderFeatures(resourceProviderNamespace: resourceProviderNamespace, listAvailable: listAvailable, featureName: null);
+        }
+
+        /// <summary>
+        /// Lists the features that ARM knows about
+        /// </summary>
+        /// <param name="resourceProviderNamespace">When specified, returns all features that are defined by this resource provider namespace</param>
+        /// <param name="featureName">When specified, returns all features that have this name</param>
+        /// <param name="listAvailable">When set to true, lists all features that are available including those not registered on the current subscription</param>
+        private PSProviderFeature[] ListPSProviderFeatures(string resourceProviderNamespace = null, string featureName = null, bool listAvailable = false)
```

Comment:
Nit: encapsulate the passed parameters into a class `FilterPSProviderOptions` and check on set properties inside this method instead. Search for `FilterAuthorizationRules` as an example